jaxopt/_src/levenberg_marquardt.py

```diff
@@ -0,0 +1,372 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Levenberg-Marquardt algorithm in JAX."""
+
+from typing import Any
+from typing import Callable
+from typing import NamedTuple
+from typing import Optional
+from typing import Union
+
+from dataclasses import dataclass
+
+from typing_extensions import Literal
+import jax
+import jax.numpy as jnp
+
+from jaxopt._src import base
+from jaxopt._src import linear_solve
+from jaxopt._src.linear_solve import solve_cg
+from jaxopt._src.linear_solve import solve_bicgstab
+from jaxopt._src.linear_solve import solve_gmres
+from jaxopt._src.linear_solve import solve_cholesky
+from jaxopt._src.linear_solve import solve_inv
+from jaxopt._src.tree_util import tree_l2_norm, tree_inf_norm, tree_sub, tree_add, tree_mul
+from jax.experimental import host_callback as hcb
+
+
+class LevenbergMarquardtState(NamedTuple):
+  """Named tuple containing state information."""
+  iter_num: int
+  damping_factor: float
+  increase_factor: float
+  residual: Any
+  loss: Any
+  delta: Any
+  error: float
+  gradient: Any
+  aux: Optional[Any] = None
+
+
+@dataclass(eq=False)
+class LevenbergMarquardt(base.IterativeSolver):
+  """Levenberg-Marquardt nonlinear least-squares solver.
+
+    Given the residual function `func` (x): R^m -> R^n, `least_squares` finds a
+    local minimum of the cost function F(x):
+
+    ```
+    argmin_x F(x) = 0.5 * sum(f_i(x)**2), i = 0, ..., m - 1
+    f(x) = func(x, *args)
+    ```
+
+    The convergence is achieved once the coeff update satisfies the criteria:
+    ```
+    ||dcoeffs||_2 <= xtol * (||coeffs||_2 + xtol)
+    ```
+    or the gradient satisfies ``||grad(f)||_inf <= gtol``.
+
+  Attributes:
+    residual_fun: a smooth function of the form ``residual_fun(x, *args,
+      **kwargs)``.
+    maxiter: maximum increase_factormber of iterations.
+    damping_parameter: The parameter which adds a correction to the equation
+      derived for updating the coefficients using Gauss-Newton method. Please
+      see section 3.2. of K. Madsen et al. in the book "Methods for nonlinear
+      least squares problems" for more information.
+    stop_criteria: The criteria to use for the convergence of the while loop.
```

Comment:
I would call it `stopping_criterion` (`criteria` is plural)

Reply:
Done.